mobile/src/main/java/ch/epfl/sdp/mobile/ui/play/ExpandableFloatingActionButton.kt

```diff
@@ -0,0 +1,240 @@
+      modifier = modifier,
+  ) {
+    AnimatedContent(
+        targetState = state.expanded,
+        contentAlignment = Alignment.Center,
+        transitionSpec = { FadeEnterTransition with FadeExitTransition },
+    ) { expanded ->
+      ProvideTextStyle(MaterialTheme.typography.button) {
+        if (expanded) ExpandedContent(content = expandedContent)
+        else
+            CollapsedContent(
+                modifier =
+                    Modifier.clickable { state.expanded = true }
+                        .padding(vertical = 20.dp, horizontal = 24.dp),
+                content = collapsedContent,
+            )
+      }
+    }
+  }
+}
+
+/**
+ * An item that may be displayed within an [ExpandableFloatingActionButton] if it's currently
+ * expanded. These items are clickable and essentially work like buttons.
+ *
+ * @param onClick the callback which will be called when the button is clicked.
+ * @param icon the slot for the action icon.
+ * @param modifier the [Modifier] for this composable.
+ * @param text the slot for the action text.
+ */
+@Composable
+fun ExpandableFloatingActionButtonItem(
+    onClick: () -> Unit,
+    icon: @Composable () -> Unit,
+    modifier: Modifier = Modifier,
+    text: @Composable () -> Unit,
+) {
+  Row(
+      modifier = modifier.clickable { onClick() }.padding(vertical = 20.dp, horizontal = 24.dp),
+      horizontalArrangement = Arrangement.SpaceBetween,
+      verticalAlignment = Alignment.CenterVertically,
+  ) {
+    // This box is required for intrinsic measurements to work properly, since icons don't quite
+    // support them properly yet.
+    Box(Modifier.size(24.dp), Alignment.Center) { icon() }
+    Spacer(Modifier.width(16.dp))
+    text()
+  }
+}
+
+// Content wrappers with proper Arrangement and Alignment.
+
+@Composable
+private fun CollapsedContent(
+    modifier: Modifier = Modifier,
+    content: @Composable RowScope.() -> Unit,
+) = Row(modifier, Arrangement.Start, Alignment.CenterVertically, content)
+
+@Composable
+private fun ExpandedContent(
+    modifier: Modifier = Modifier,
+    content: @Composable ColumnScope.() -> Unit,
+) = Column(modifier.width(IntrinsicSize.Max), Arrangement.Top, Alignment.Start, content)
+
+// State management.
+
+/**
+ * Creates and remembers an [ExpandableFloatingActionButtonState].
+ *
+ * @param initiallyExpanded true if the floating action button should initially be expanded.
+ */
+@Composable
+fun rememberExpandableFloatingActionButtonState(
+    initiallyExpanded: Boolean = false,
+): ExpandableFloatingActionButtonState {
+  val expanded = rememberSaveable { mutableStateOf(initiallyExpanded) }
+  return SnapshotExpandableFloatingActionButtonState(expanded)
+}
+
+/**
+ * An implementation of an [ExpandableFloatingActionButtonState] that delegates to a backing
+ * [MutableState].
+ *
+ * @param state the [MutableState] to which the expanded state is delegated.
+ */
+private class SnapshotExpandableFloatingActionButtonState(
```

Comment:
[Question] Why do we chose to delegate expanded state to a `MutableState` ? Why simply not make it itself `MutableState` then use the `by` operand ?